### PR TITLE
Add a generator

### PR DESCRIPTION
**Pull request by danieldk**
## Description

This change adds a very basic generator. It only does argmax decoding, no sampling, no beam search, no logit temperatures, no stopping conditions other than the EOS piece, etc. The goal is to put something in place that we can iterate on. The main classes added in this change are:

- `Generator` provides the generation loop.
- `StringGenerator` is a wrapper for `Generator` that takes pieces and produces pieces.
- `DollyV2Generator` uses these two classes and wraps user prompts in model-specific instructions.

There are many other changes throughout the code. The largest is that we introduce a generic type bound by `CacheProtocol`, so that we can treat cache (such as the key-value cache) in a generic way.

This change also adds a `veryslow` test mark for tests that we probably don't want to run in CI because the models are too large and/or slow, such as the Dolly V2 generation test.

### Types of change

Feature

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: curated_transformers/generation/string_generator.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        See the :meth:`.generate` method.
+        """
+        return self.generate(prompts, eos_id)
+
+    def generate(
+        self, prompts: Iterable[InputChunks], eos_id: int
+    ) -> Iterator[List[Tuple[int, str]]]:
+        """
+        Generate text using the given prompts. This function yields for
+        each generation step a list of requence identifiers and the
```
Comment (shadeMe) [inline curated_transformers/generation/string_generator.py:43]:
```suggestion
        each generation step a list of sequence identifiers and the
```

File: curated_transformers/generation/string_generator.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+        device = next(self.inner.model.parameters()).device
+        pieces = self.tokenizer(prompts)
+        ids = pieces.padded_tensor(padding_id=0, pad_left=True).to(device)
+        attention_mask = pieces.attention_mask(pad_left=True).to(device)
+
+        for seq_ids, outputs in self.inner(
+            ids=ids, attention_mask=attention_mask, eos_id=eos_id
+        ):
+            decoded = self.tokenizer.decode(outputs.tolist())
```
Comment (shadeMe) [inline curated_transformers/generation/string_generator.py:64]: We need to handle non-EOS special tokens here to prevent `KeyErrors`.

File: curated_transformers/models/attention.py [hunk 3]
```diff
@@ -70,13 +70,48 @@ def apply_causal_mask(input: Tensor) -> Tensor:
...
+
+class CacheProtocol(Protocol):
+    def filter_batch_items(self: CacheProtocolSelf, mask: Tensor) -> CacheProtocolSelf:
+        """
+        Filter batch sequences from the cache.
+
+        Sequences for which the mask is ``True`` are retained.
+
+        :param mask:
+            Mask of batch items to retain.
```
Comment (shadeMe) [inline curated_transformers/models/attention.py:84]:
```suggestion
            Mask of batch items to retain.
            **Shape:** (batch,)
```

File: curated_transformers/generation/state.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+from torch import Tensor
+
+from ..models.attention import CacheT
+
+
+class GeneratorState(Generic[CacheT]):
+    """Generator state."""
+
+    attention_mask: Tensor
+    cache: Optional[List[CacheT]]
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/state.py:12]: I'd like to make it non-optional, but the problem is that we don't have a cache yet when decoding the prompt.

File: curated_transformers/generation/string_generator.py (hunk 2 above)
Comment (danieldk) [reply curated_transformers/generation/string_generator.py:64]: It needs to go into the model. Either as a logit processor to mask them out (so that they don't get predicted) or as a stopping condition. But the abstractions that are needed are not in the basic generator yet (the scaffolding for logit processing will be the next PR).

File: curated_transformers/generation/models/__init__.py [hunk 5]
```diff
@@ -0,0 +1,2 @@
+from .dolly_v2 import DollyV2Generator
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/models/__init__.py]: Changed the naming as discussed.

File: curated_transformers/generation/models/dolly_v2.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            _to_prompt_with_instructions(prompt) for prompt in prompts
+        ]
+        yield from self.generator(prompts_with_instructions, eos_id=self.eos_id)
+
+
+def _to_prompt_with_instructions(prompt):
+    return InputChunks(
+        [
+            TextChunk(INTRO_BLURB),
+            SpecialPieceChunk(INSTRUCTION_KEY, before="\n\n", after="\n"),
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/models/dolly_v2.py]: Good idea, added.

File: curated_transformers/generation/models/generation_model.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+
+# Only provided as typing.Self in Python 3.11+.
+Self = TypeVar("Self", bound="GenerationModel")
+
+
+class GenerationModel(ABC):
+    @classmethod
+    @abstractmethod
+    def from_hf_hub(
```
[the remark this answers is not in the document]
Comment (danieldk) [reply curated_transformers/generation/models/generation_model.py]: That has `Module`-specific stuff, but I made it a separate mixin.

File: curated_transformers/generation/state.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shadeMe) [reply curated_transformers/generation/state.py:12]: Fair point.

**Review by shadeMe — APPROVED** [review]
